Cannot restart after shutdown / Ctrl+C (etcdmain: database file does not match with snapshot)

I have a cluster of 3 and I'm testing failover scenarios right now. I simulate power outages (shutting of a VM) and service failures (Ctrl+C on etcd). As of now, shutting off an instance has terrible consequences, as it cannot recover by simply rebooting a machine / restarting the service:

```
$ etcd
// ...runs...
<Press Ctrl+C>

$ etcd
2016-07-03 02:05:21.753663 I | flags: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://rly4:2379
2016-07-03 02:05:21.754254 I | flags: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd/data
2016-07-03 02:05:21.754550 I | flags: recognized and used environment variable ETCD_DEBUG=true
2016-07-03 02:05:21.754805 I | flags: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://rly4:2380
2016-07-03 02:05:21.755066 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER=rly4=http://rly4:2380,rly1=http://rly1:2380,rly5=http://rly5:2380
2016-07-03 02:05:21.755317 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER_STATE=existing
2016-07-03 02:05:21.755585 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER_TOKEN=rly-cluster-1
2016-07-03 02:05:21.755833 I | flags: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://rly4:2379,http://127.0.0.1:2379
2016-07-03 02:05:21.756038 I | flags: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://rly4:2380
2016-07-03 02:05:21.756276 I | flags: recognized and used environment variable ETCD_NAME=rly4
2016-07-03 02:05:21.756525 I | flags: recognized and used environment variable ETCD_WAL_DIR=/var/lib/etcd/wal
2016-07-03 02:05:21.756784 I | etcdmain: etcd Version: 3.0.0
2016-07-03 02:05:21.757038 I | etcdmain: Git SHA: 6f48bda
2016-07-03 02:05:21.757261 I | etcdmain: Go Version: go1.6.2
2016-07-03 02:05:21.757478 I | etcdmain: Go OS/Arch: linux/amd64
2016-07-03 02:05:21.757696 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
2016-07-03 02:05:21.757950 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-07-03 02:05:21.761023 I | etcdmain: listening for peers on http://rly4:2380
2016-07-03 02:05:21.761065 I | etcdmain: listening for client requests on 127.0.0.1:2379
2016-07-03 02:05:21.763682 I | etcdmain: listening for client requests on rly4:2379
2016-07-03 02:05:21.804445 I | etcdserver: recovered store from snapshot at index 141365
2016-07-03 02:05:21.804844 I | etcdserver: name = rly4
2016-07-03 02:05:21.805118 I | etcdserver: data dir = /var/lib/etcd/data
2016-07-03 02:05:21.805389 I | etcdserver: member dir = /var/lib/etcd/data/member
2016-07-03 02:05:21.805661 I | etcdserver: dedicated WAL dir = /var/lib/etcd/wal
2016-07-03 02:05:21.805891 I | etcdserver: heartbeat = 100ms
2016-07-03 02:05:21.806169 I | etcdserver: election = 1000ms
2016-07-03 02:05:21.806428 I | etcdserver: snapshot count = 10000
2016-07-03 02:05:21.806696 I | etcdserver: advertise client URLs = http://rly4:2379
2016-07-03 02:05:21.807377 I | etcdserver: restarting member 89889c482441ec6e in cluster d0ce954dc5f082d0 at commit index 141549
2016-07-03 02:05:21.807753 I | raft: 89889c482441ec6e became follower at term 1966
2016-07-03 02:05:21.808041 I | raft: newRaft 89889c482441ec6e [peers: [89889c482441ec6e,d27323ab9b295f50,e790e6b697b3c219], term: 1966, commit: 141549, applied: 141365, lastindex: 141549, lastterm: 1966]
2016-07-03 02:05:21.808406 I | membership: added member d27323ab9b295f50 [http://rly5:2380] to cluster d0ce954dc5f082d0 from store
2016-07-03 02:05:21.808751 I | membership: added member e790e6b697b3c219 [http://rly1:2380] to cluster d0ce954dc5f082d0 from store
2016-07-03 02:05:21.809021 I | membership: added member 89889c482441ec6e [http://rly4:2380] to cluster d0ce954dc5f082d0 from store
2016-07-03 02:05:21.809295 I | membership: set the cluster version to 3.0 from store
2016-07-03 02:05:21.818080 I | etcdmain: stopping listening for client requests on rly4:2379
2016-07-03 02:05:21.818511 I | etcdmain: stopping listening for client requests on rly4:2379
2016-07-03 02:05:21.818707 I | etcdmain: stopping listening for peers on http://rly4:2380
2016-07-03 02:05:21.818894 C | etcdmain: database file (/var/lib/etcd/data/member/snap/db index 140014) does not match with snapshot (index 141365).

$ find
./data
./data/member
./data/member/snap
./data/member/snap/00000000000007ae-0000000000022835.snap
./data/member/snap/db
./wal
./wal/0000000000000000-0000000000000000.wal
./wal/0.tmp
```

The only way to recover from this right now is to delete the data+wal dir, remove the member, re-add it and then restart the daemon. However, I fear that if all fail at the same time (power failure), I will lose all data.

Related: After removing the data+wal dir, restarting (without removing the member first) prints a stack trace:

```
# etcd
2016-07-03 02:20:28.031273 I | flags: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://rly4:2379
2016-07-03 02:20:28.031752 I | flags: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd/data
2016-07-03 02:20:28.032234 I | flags: recognized and used environment variable ETCD_DEBUG=true
2016-07-03 02:20:28.032705 I | flags: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://rly4:2380
2016-07-03 02:20:28.033143 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER=rly4=http://rly4:2380,rly1=http://rly1:2380,rly5=http://rly5:2380
2016-07-03 02:20:28.033508 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER_STATE=existing
2016-07-03 02:20:28.035314 I | flags: recognized and used environment variable ETCD_INITIAL_CLUSTER_TOKEN=rly-cluster-1
2016-07-03 02:20:28.035679 I | flags: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://rly4:2379,http://127.0.0.1:2379
2016-07-03 02:20:28.035896 I | flags: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://rly4:2380
2016-07-03 02:20:28.036123 I | flags: recognized and used environment variable ETCD_NAME=rly4
2016-07-03 02:20:28.036379 I | flags: recognized and used environment variable ETCD_WAL_DIR=/var/lib/etcd/wal
2016-07-03 02:20:28.036688 I | etcdmain: etcd Version: 3.0.0
2016-07-03 02:20:28.036923 I | etcdmain: Git SHA: 6f48bda
2016-07-03 02:20:28.037145 I | etcdmain: Go Version: go1.6.2
2016-07-03 02:20:28.037365 I | etcdmain: Go OS/Arch: linux/amd64
2016-07-03 02:20:28.037586 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
2016-07-03 02:20:28.040920 I | etcdmain: listening for peers on http://rly4:2380
2016-07-03 02:20:28.040959 I | etcdmain: listening for client requests on 127.0.0.1:2379
2016-07-03 02:20:28.043601 I | etcdmain: listening for client requests on rly4:2379
2016-07-03 02:20:28.057481 I | netutil: resolving rly1:2380 to 192.168.0.31:2380
2016-07-03 02:20:28.061152 I | netutil: resolving rly1:2380 to 192.168.0.31:2380
2016-07-03 02:20:28.064142 I | netutil: resolving rly4:2380 to 192.168.0.34:2380
2016-07-03 02:20:28.067372 I | netutil: resolving rly4:2380 to 192.168.0.34:2380
2016-07-03 02:20:28.070151 I | netutil: resolving rly5:2380 to 192.168.0.35:2380
2016-07-03 02:20:28.072927 I | netutil: resolving rly5:2380 to 192.168.0.35:2380
2016-07-03 02:20:28.077895 I | etcdserver: name = rly4
2016-07-03 02:20:28.078413 I | etcdserver: data dir = /var/lib/etcd/data
2016-07-03 02:20:28.078798 I | etcdserver: member dir = /var/lib/etcd/data/member
2016-07-03 02:20:28.079151 I | etcdserver: dedicated WAL dir = /var/lib/etcd/wal
2016-07-03 02:20:28.079490 I | etcdserver: heartbeat = 100ms
2016-07-03 02:20:28.079506 I | etcdserver: election = 1000ms
2016-07-03 02:20:28.079514 I | etcdserver: snapshot count = 10000
2016-07-03 02:20:28.079527 I | etcdserver: advertise client URLs = http://rly4:2379
2016-07-03 02:20:28.083747 I | etcdserver: starting member 89889c482441ec6e in cluster d0ce954dc5f082d0
2016-07-03 02:20:28.084129 I | raft: 89889c482441ec6e became follower at term 0
2016-07-03 02:20:28.084149 I | raft: newRaft 89889c482441ec6e [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-07-03 02:20:28.084157 I | raft: 89889c482441ec6e became follower at term 1
2016-07-03 02:20:28.097873 I | rafthttp: started HTTP pipelining with peer d27323ab9b295f50
2016-07-03 02:20:28.098338 I | rafthttp: started HTTP pipelining with peer e790e6b697b3c219
2016-07-03 02:20:28.098619 I | rafthttp: starting peer d27323ab9b295f50...
2016-07-03 02:20:28.099148 I | rafthttp: started HTTP pipelining with peer d27323ab9b295f50
2016-07-03 02:20:28.099685 I | rafthttp: started streaming with peer d27323ab9b295f50 (writer)
2016-07-03 02:20:28.105343 I | rafthttp: started peer d27323ab9b295f50
2016-07-03 02:20:28.105805 I | rafthttp: added peer d27323ab9b295f50
2016-07-03 02:20:28.106256 I | rafthttp: starting peer e790e6b697b3c219...
2016-07-03 02:20:28.106673 I | rafthttp: started HTTP pipelining with peer e790e6b697b3c219
2016-07-03 02:20:28.114568 I | rafthttp: started streaming with peer d27323ab9b295f50 (stream Message reader)
2016-07-03 02:20:28.115871 I | rafthttp: started streaming with peer d27323ab9b295f50 (writer)
2016-07-03 02:20:28.116274 I | rafthttp: started streaming with peer d27323ab9b295f50 (stream MsgApp v2 reader)
2016-07-03 02:20:28.116737 I | rafthttp: started peer e790e6b697b3c219
2016-07-03 02:20:28.117054 I | rafthttp: added peer e790e6b697b3c219
2016-07-03 02:20:28.117337 I | etcdserver: starting server... [version: 3.0.0, cluster version: to_be_decided]
2016-07-03 02:20:28.117781 I | rafthttp: started streaming with peer e790e6b697b3c219 (writer)
2016-07-03 02:20:28.118010 I | rafthttp: started streaming with peer e790e6b697b3c219 (writer)
2016-07-03 02:20:28.129196 I | rafthttp: started streaming with peer e790e6b697b3c219 (stream Message reader)
2016-07-03 02:20:28.131685 I | rafthttp: started streaming with peer e790e6b697b3c219 (stream MsgApp v2 reader)
2016-07-03 02:20:28.132625 I | raft: 89889c482441ec6e [term: 1] received a MsgHeartbeat message with higher term from e790e6b697b3c219 [term: 1966]
2016-07-03 02:20:28.133191 I | raft: 89889c482441ec6e became follower at term 1966
2016-07-03 02:20:28.133723 C | raft: tocommit(141549) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(141549) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 91 [running]:
panic(0xd44e00, 0xc8201e4680)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/coreos/etcd/cmd/vendor/github.com/coreos/pkg/capnslog.(*PackageLogger).Panicf(0xc820192780, 0x12d2e00, 0x5d, 0xc8201d8880, 0x2, 0x2)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/pkg/capnslog/pkg_logger.go:75 +0x191
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc820200460, 0x228ed)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/log.go:191 +0x1a6
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc8201d52b0, 0x8, 0x89889c482441ec6e, 0xe790e6b697b3c219, 0x7ae, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:844 +0x44
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.stepFollower(0xc8201d52b0, 0x8, 0x89889c482441ec6e, 0xe790e6b697b3c219, 0x7ae, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:806 +0x1194
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc8201d52b0, 0x8, 0x89889c482441ec6e, 0xe790e6b697b3c219, 0x7ae, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/raft.go:596 +0x66d
github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.(*node).run(0xc8201b5da0, 0xc8201d52b0)
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/node.go:317 +0x8ce
created by github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft.StartNode
	/home/gyuho/go/src/github.com/coreos/etcd/cmd/vendor/github.com/coreos/etcd/raft/node.go:204 +0x777
```

This is really hard. Do you have instructions how to set this stuff up from scratch? I feel I'm randomly poking around and I'm not getting anywhere. I managed to put things in the right place, but now when I run `go get` or `go get ./...` (I googled this, no idea what it does), I get this:

```
7/gopath/src/github.com/coreos/etcd:0a9d1c3 ✓ ❯❯❯ go get ./...
# golang.org/x/net/http2/hpack
../../../golang.org/x/net/http2/hpack/huffman.go:14: undefined: sync.Pool
# golang.org/x/sys/unix
../../../golang.org/x/sys/unix/syscall_solaris.go:38: clen redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:817
../../../golang.org/x/sys/unix/syscall_solaris.go:51: ParseDirent redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:830
../../../golang.org/x/sys/unix/syscall_solaris.go:77: Pipe redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux_amd64.go:110
../../../golang.org/x/sys/unix/syscall_solaris.go:89: (*SockaddrInet4).sockaddr redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:311
../../../golang.org/x/sys/unix/syscall_solaris.go:103: (*SockaddrInet6).sockaddr redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:325
../../../golang.org/x/sys/unix/syscall_solaris.go:118: (*SockaddrUnix).sockaddr redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:340
../../../golang.org/x/sys/unix/syscall_solaris.go:144: Getsockname redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:525
../../../golang.org/x/sys/unix/syscall_solaris.go:153: ImplementsGetwd redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:184
../../../golang.org/x/sys/unix/syscall_solaris.go:157: Getwd redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:188
../../../golang.org/x/sys/unix/syscall_solaris.go:178: Getgroups redeclared in this block
	previous declaration at ../../../golang.org/x/sys/unix/syscall_linux.go:201
../../../golang.org/x/sys/unix/syscall_solaris.go:178: too many errors
# github.com/coreos/etcd/pkg/httputil
pkg/httputil/httputil.go:18: req.Cancel undefined (type *http.Request has no field or method Cancel)
# github.com/coreos/etcd/pkg/fileutil
pkg/fileutil/lock_linux.go:37: undefined: syscall.Flock_t
pkg/fileutil/lock_linux.go:50: undefined: syscall.Flock_t
pkg/fileutil/lock_linux.go:51: undefined: syscall.FcntlFlock
pkg/fileutil/lock_linux.go:68: undefined: syscall.FcntlFlock
pkg/fileutil/lock_linux.go:89: undefined: syscall.FcntlFlock
# github.com/coreos/etcd/cmd/vendor/github.com/prometheus/common/model
cmd/vendor/github.com/prometheus/common/model/signature.go:33: undefined: sync.Pool
# github.com/coreos/etcd/cmd/vendor/golang.org/x/net/http2/hpack
cmd/vendor/golang.org/x/net/http2/hpack/huffman.go:14: undefined: sync.Pool
# github.com/coreos/etcd/cmd/vendor/github.com/akrennmair/gopcap
cmd/vendor/github.com/akrennmair/gopcap/pcap.go:12:18: fatal error: pcap.h: No such file or directory
 #include <pcap.h>
                  ^
compilation terminated.
# github.com/coreos/etcd/cmd/vendor/golang.org/x/sys/unix
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:37: clen redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:784
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:50: ParseDirent redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:797
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:76: Pipe redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux_amd64.go:108
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:88: (*SockaddrInet4).sockaddr redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:291
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:102: (*SockaddrInet6).sockaddr redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:305
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:117: (*SockaddrUnix).sockaddr redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:320
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:141: Getsockname redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:492
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:153: ImplementsGetwd redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:164
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:155: Getwd redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:168
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:164: Getgroups redeclared in this block
	previous declaration at cmd/vendor/golang.org/x/sys/unix/syscall_linux.go:181
cmd/vendor/golang.org/x/sys/unix/syscall_solaris.go:164: too many errors
# github.com/akrennmair/gopcap
../../akrennmair/gopcap/pcap.go:12:18: fatal error: pcap.h: No such file or directory
 #include <pcap.h>
                  ^
compilation terminated.
# github.com/ugorji/go/codec
../../ugorji/go/codec/encode.go:1377: undefined: sync.Pool
../../ugorji/go/codec/encode.go:1387: undefined: sync.Pool
../../ugorji/go/codec/helper.go:247: undefined: sync.Pool
# github.com/coreos/etcd/cmd/vendor/github.com/ugorji/go/codec
cmd/vendor/github.com/ugorji/go/codec/encode.go:1341: undefined: sync.Pool
cmd/vendor/github.com/ugorji/go/codec/encode.go:1351: undefined: sync.Pool
```

--- a/etcdserver/server.go
+++ b/etcdserver/server.go
@@ -1170,8 +1170,7 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
 			}
 			plog.Panicf("unexpected create snapshot error %v", err)
 		}
-		// commit v3 storage because WAL file before snapshot index
-		// could be removed after SaveSnap.
+		// commit kv to write metadata (for example: consistent index) to disk.
 		s.KV().Commit()
 		// SaveSnap saves the snapshot and releases the locked wal files
 		// to the snapshot index.

--- a/etcdserver/snapshot_merge.go
+++ b/etcdserver/snapshot_merge.go
@@ -39,6 +39,8 @@ func (s *EtcdServer) createMergedSnapshotMessage(m raftpb.Message, snapi uint64,
 		plog.Panicf("store save should never fail: %v", err)
 	}
 
+	// commit kv to write metadata(for example: consistent index).
+	s.KV().Commit()
 	dbsnap := s.be.Snapshot()
 	// get a snapshot of v3 KV as readCloser
 	rc := newSnapshotReaderCloser(dbsnap)

